runtime/ms-rest-azure/lib/credentials/msiTokenCredentials.js

```diff
@@ -36,16 +36,55 @@ class MSITokenCredentials {
   }
 
   /**
-   * Parses a tokenResponse json string into a object, and converts properties on the first level to camelCase
+   * Parses a tokenResponse json string into a object, and converts properties on the first level to camelCase.
+   * This method tries to standardize the tokenResponse
    * @param  {string} body  A json string
    * @return {object} [tokenResponse] The tokenResponse (tokenType and accessToken are the two important properties). 
    */
   parseTokenResponse(body) {
+    // Docs show different examples of possible MSI responses for different services. https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview
+    // expires_on - is a Date like string in this doc
+    //   - https://docs.microsoft.com/en-us/azure/app-service/app-service-managed-service-identity#rest-protocol-examples
+    // In other doc it is stringified number.
+    //   - https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/tutorial-linux-vm-access-arm#get-an-access-token-using-the-vms-identity-and-use-it-to-call-resource-manager
     let parsedBody = JSON.parse(body);
     parsedBody.accessToken = parsedBody['access_token'];
-    parsedBody.tokenType = parsedBody['token_type'];
     delete parsedBody['access_token'];
+    parsedBody.tokenType = parsedBody['token_type'];
     delete parsedBody['token_type'];
+    if (parsedBody['refresh_token']) {
+      parsedBody.refreshToken = parsedBody['refresh_token'];
+      delete parsedBody['refresh_token'];
+    }
+    if (parsedBody['expires_in']) {
+      parsedBody.expiresIn = parsedBody['expires_in'];
+      if (typeof parsedBody['expires_in'] === 'string') {
+        // normal number as a string '1504130527'
+        parsedBody.expiresIn = parseInt(parsedBody['expires_in'], 10);
+      }
+      delete parsedBody['expires_in'];
+    }
+    if (parsedBody['not_before']) {
+      parsedBody.notBefore = parsedBody['not_before'];
+      if (typeof parsedBody['not_before'] === 'string') {
+        // normal number as a string '1504130527'
+        parsedBody.notBefore = parseInt(parsedBody['not_before'], 10);
+      }
+      delete parsedBody['not_before'];
+    }
+    if (parsedBody['expires_on']) {
+      parsedBody.expiresOn = parsedBody['expires_on'];
+      if (typeof parsedBody['expires_on'] === 'string') {
+        // possibly a Date string '09/14/2017 00:00:00 PM +00:00'
+        if (parsedBody['expires_on'].includes(':') || parsedBody['expires_on'].includes('/')) {
+          parsedBody.expiresOn = new Date(parsedBody['expires_on'], 10);
+        } else {
+          // normal number as a string '1504130527'
+          parsedBody.expiresOn = new Date(parseInt(parsedBody['expires_on'], 10));
+        }
+      }
+      delete parsedBody['expires_on'];
+    }

```

Comment:
If different services return different MSI responses, wouldn't it be preferable to address each response in the corresponding MSICredentials?
i.e. define seperate `parseTokenResponse()` for each MSI response.

Seems to me like trying to accommodate the different logic for every msi response (and new ones in the future) will be harder to maintain.

Reply:
After our talk, I'm okay with this for now given we only have to deal with two cases.
Depending on the new MSI responses in the future, we would ideally separate the common logic and the service specific logic in the corresponding parent and children classes.